### ISSUE TITLE
Remove keycodes?

### ISSUE THREAD

**Issue by slouken** [libsdl-org/SDL#8636]
It has been a continual source of confusion that there are scancodes, keycodes, and text input events.

Scancodes are used for positional input, e.g. WASD movement even though someone might be using a DVORAK layout.
Keycodes are the label on the key referenced by a scancode
Text input is the composed input based on the current keyboard state.

People often try to use keycodes as text input, manually applying shift modifiers and so forth. We clearly need a way to map from scancode to key label for UI purposes, but can someone remind me why we need to have that in the keyboard event?

@libsdl-org/a-team?

**Comment by slouken** [comment]
@DJm00n, thoughts?

**Comment by DJm00n** [comment]
In our game engine we only have scancodes that represents the same button regardless keyboard layout. And also `GetButtonName(scanCode)` API for GUI use cases that returns buttons names according to the current active keyboard layout or default name (if we cannot get proper mappings on this platform). Thats all.

I think scancode-central approach is mandatory in gamedev to be consistent between keyboard layouts (QWERTY/AZERTY/QWERTZ/DVORAK etc etc).

As for me `SDL_Keycodes` and VK codes that are used in Windows is something like "input action" concept that is usually used in game engines (her is such [example from UE](https://docs.unrealengine.com/4.26/en-US/InteractiveExperiences/Input/EnhancedInput/) or [from Unity](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.0/manual/Actions.html#creating-actions)) but designed for simple Win32 GUI use case, not games. Generally each game engine have to implement its own "input action" and "input action remapping" system.

Long story short - I think `SDL_Keycodes` is redundant in SDL. `SDL_Scancode`, `SDL_GetScancodeName()` (for "serialize to text config file" etc use cases) and `GetScancodeDisplayName()` (for show in game GUI use cases) is enough for most uses.

**Comment by Kontrabant** [comment]
Remapping keys at the system level changes the mapped key codes, but not the scan codes. For example, depending on the input locale, the right alt key can emit either right alt or altgr/mode/level-3-shift, and doing something like swapping the caps lock and escape keys (apparently very common among vim users) doesn't change the scan codes emitted by the keys, but the key codes they map to are swapped.

The key code is basically required if you want to know what the expected behavior of the pressed key is, vs just knowing its location on the keyboard.

**Comment by slime73** [comment]
To add to that, regardless of whether keycodes are part of key input events or not, I think being able to query the scancode given the name of a key as it appears to the user or is configured by the user is still pretty useful, alongside the reverse.

**Comment by icculus** [comment]
Also our own test apps have a bunch of "press 'f' to toggle fullscreen"

And we really _do_ want the Escape key to cancel things no matter where it is on the keyboard, etc.

**Comment by slouken** [comment]
Okay, now's the time for silly ideas, and this one falls in that bucket. Thanks everyone!

**Comment by DJm00n** [comment]
@icculus 
>Also our own test apps have a bunch of "press 'f' to toggle fullscreen"

The thing is that not all keyboards even contain 'f' key. In Windows A..Z VK codes are always present on keyboard (even in [Russian](https://kbdlayout.info/kbdru) or [Gothic](https://kbdlayout.info/kbdgthc) keyboards) but this is not case of `SDL_Keycode` now (at least on Win32), after my change https://github.com/libsdl-org/SDL/commit/7e86b6aef2b16e08a21a03948b6e199e9b4a1ffa - right now they are mapped to the actual Unicode symbol that is printer on a key press as stated in comment near `SDL_Keycode` declaration:

```cpp
 *  The SDL virtual key representation.
 *
 *  Values of this type are used to represent keyboard keys using the current
 *  layout of the keyboard.  These values include Unicode values representing
 *  the unmodified character that would be generated by pressing the key, or
 *  an SDLK_* constant for those keys that do not generate characters.
```

Should I uncomment VK A..Z part in `WIN_UpdateKeymap`?

**Comment by DJm00n** [comment]
I've given it some more thought.. and it looks like we need:
- keep `SDL_Keycode` but change the comment next to it to indicate that `SDLK_a`..`SDLK_z`  may not always represent what will be typed in current keyboard layout (`SDL_SCANCODE_Q` is mapped to `SDLK_a` in French keyboard, but to `SDLK_a` in Russian given that Russian keyboard doesn't have ASCII `a` key at all)
- uncomment the VK A..Z part in `WIN_UpdateKeymap` (see https://github.com/libsdl-org/SDL/commit/7e86b6aef2b16e08a21a03948b6e199e9b4a1ffa)
- fix `SDL_GetKeyName` - so that it returns the real name of the key in the current layout (for `SDLK_a`..`SDLK_z` - character that is printed on a key press), instead of just call tp `SDL_UCS4ToUTF8(keycode)`.For example, it should return `Й` for `SDLK_q` in [Russian keyboard layout](https://kbdlayout.info/kbdru).
